(recopilado de http://www.mclibre.org/consultar/python/lecciones/python_entrada_teclado.html)

# La función input()

La función input() permite obtener texto escrito en el teclado por el usuario. El programa, al detectar esta función, se queda esperando a que se escriba algo y se pulse la tecla Enter. Un ejemplo de esto se ve en el siguiente ejemplo:

In [1]:
print("¿Cómo se llama?")
nombre = input()
print("Me alegro de conocerle,", nombre)

¿Cómo se llama?
Usuario
Me alegro de conocerle, Usuario


En el ejemplo anterior, el usuario escribe su respuesta en una línea distinta a la pregunta porque Python añade un salto de línea al final de cada print().


&nbsp;Si se prefiere que el usuario escriba su respuesta a continuación de la pregunta, se podría utilizar el argumento opcional **end** en la función **print()**, que indica el carácter o caracteres a utilizar en vez del salto de línea. 

&nbsp;Para separar la respuesta de la pregunta se ha añadido un espacio al final de la pregunta.

In [2]:
print("¿Cómo se llama? ", end="")
nombre = input()
print("Me alegro de conocerle,", nombre)

¿Cómo se llama? Usuario
Me alegro de conocerle, Usuario


Otra solución, más compacta, es aprovechar que a la función input() se le puede enviar un argumento que se escribe en la pantalla (sin añadir un salto de línea):

In [3]:
nombre = input("¿Cómo se llama? ")
print("Me alegro de conocerle,", nombre)

¿Cómo se llama? Usuario
Me alegro de conocerle, Usuario



&nbsp;
# Conversión de tipos

De forma predeterminada, la función input() convierte la entrada en una cadena. Si se quiere que Python interprete la entrada como un número entero, se debe utilizar la función int() de la siguiente manera:

In [4]:
cantidad = int(input("Ingrese cantidad de Dolares: "))
dolar= int(input("Valor Dolar: "))
print(cantidad, "dolares son", cantidad * dolar, "pesos")

Ingrese cantidad de Dolares: 100
Valor Dolar: 680
100 dolares son 68000 pesos


Si se trabaja con valores decimales, el ejemplo anterior debe adaptrase al tipo de dato **float**:

In [6]:
cantidad = float(input("Ingrese cantidad de Dolares: "))
dolar= int(input("Valor Dolar: "))
print(cantidad, "dolares son", cantidad * dolar, "pesos")

Ingrese cantidad de Dolares: 100.5
Valor Dolar: 680
100.5 dolares son 68340.0 pesos


Si no se respetan los tipos de datos de las variables de entrada, se producen errores. Si tomamamos el primer ejemplo e ingresamos un valor con decimales, se produciria un error como el siguiente:

In [7]:
cantidad = int(input("Ingrese cantidad de Dolares: "))
dolar= int(input("Valor Dolar: "))
print(cantidad, "dolares son", cantidad * dolar, "pesos")

Ingrese cantidad de Dolares: 100.5


ValueError: invalid literal for int() with base 10: '100.5'

Esto se produce al ingresar un valor con decimales **(100.5)** teniendo una conversión de tipo **int**

Esto no ocurre al ingresar un entero en una variable definida para recibir un valor real (con decimales). Lo único es que se representará con un decimal

In [8]:
peso = float(input("Ingrese su peso en kg: "))
print("Su peso es", peso, "kg")

Ingrese su peso en kg: 80
Su peso es 80.0 kg


# Variables como argumento de la función input()

La función input() sólo puede tener un argumento.

Esto puede causar problemas como en el ejemplo siguiente, en el que se quiere mostrar la respuesta de la primera instrucción en la segunda pregunta:

In [9]:
nombre = input("Dígame su nombre: ")
apellido = input("Dígame su apellido, ", nombre, ": ")
print("Me alegro de conocerle,", nombre, apellido)

Dígame su nombre: Juan


TypeError: raw_input() takes from 1 to 2 positional arguments but 4 were given

Este problema se puede resolver de varias maneras:
* Una solución consiste en separar la pregunta de la recogida de la respuesta, escribiendo dos instrucciones:

In [10]:
nombre = input("Dígame su nombre: ")
print("Dígame su apellido,", nombre, ": ", end="")
apellido = input()
print("Me alegro de conocerle,", nombre, apellido)

Dígame su nombre: James
Dígame su apellido, James : Bond
Me alegro de conocerle, James Bond


Si se quiere que los dos puntos **(:)** salgan pegados al nombre habría que utilizar la concatenación (operador +):

In [11]:
nombre = input("Dígame su nombre: ")
print("Dígame su apellido,", nombre + ": ", end="")
apellido = input()
print("Me alegro de conocerle,", nombre, apellido)

Dígame su nombre: James
Dígame su apellido, James: Bond
Me alegro de conocerle, James Bond


* Otra solución que permite incluir la pregunta en la recogida de la respuesta es utilizar la concatenación (operador +):

In [12]:
nombre = input("Dígame su nombre: ")
apellido = input("Dígame su apellido, " + nombre +  ": ")
print("Me alegro de conocerle,", nombre, apellido)

Dígame su nombre: Super
Dígame su apellido, Super: Man
Me alegro de conocerle, Super Man


Si se opta por esta solución, hay que tener en cuenta que, en caso de querer concatenar cadenas y números, se debe utilizar la función **str()**, como se ha comentado en el apartado anterior:

* Código con Error

In [13]:
numero1 = int(input("Dígame un número: "))
numero2 = int(input("Dígame un número mayor que " + numero1 +  ": "))
print("La diferencia entre ellos es", numero2 - numero1)

Dígame un número: 10


TypeError: must be str, not int

El error se debe a que el comando **+** solo se puede ocupar para concatenar cadenas de texto y **numero1** es una variable de tipo entera.

* Código corregido

In [14]:
numero1 = int(input("Dígame un número: "))
numero2 = int(input("Dígame un número mayor que " + str(numero1) +  ": "))
print("La diferencia entre ellos es", numero2 - numero1)

Dígame un número: 10
Dígame un número mayor que 10: 20
La diferencia entre ellos es 10


* Otra solución es utilizar las cadenas **" f "** introducidas en Python 3.6:

In [15]:
nombre = input("Dígame su nombre: ")
apellido = input(f"Dígame su apellido, {nombre}: ")
print(f"Me alegro de conocerle, {nombre} {apellido}")

Dígame su nombre: Elba
Dígame su apellido, Elba: Lazo
Me alegro de conocerle, Elba Lazo


In [16]:
numero1 = int(input("Dígame un número: "))
numero2 = int(input(f"Dígame un número mayor que {numero1}: "))
print(f"La diferencia entre ellos es {numero2 - numero1}")

Dígame un número: 10
Dígame un número mayor que 10: 20
La diferencia entre ellos es 10
